provider/pkg/resources/suppress_diff.go

```diff
@@ -154,36 +163,449 @@ func getTagKey(tag resource.PropertyValue) string {
+//
+// For nested paths like "fileSystemProtection/replicationOverwriteProtection", we build
+// a nested structure that JSONata can traverse:
+//
+//	{
+//	  "FileSystemProtection": {
+//	    "ReplicationOverwriteProtection": value
+//	  }
+//	}
+//
+// This allows JSONata expressions like "$uppercase(FileSystemProtection.ReplicationOverwriteProtection)"
+// to work correctly.
+func buildCfnContext(baseContext map[string]interface{}, path string, value interface{}) map[string]interface{} {
+	ctx := make(map[string]interface{})
+
+	// Copy base context (sibling properties)
+	for k, v := range baseContext {
+		ctx[k] = v
+		// Add multiple name variations for CFN compatibility
+		addCfnNameVariations(ctx, k, v)
+	}
+
+	// Build nested structure from path segments
+	segments := strings.Split(path, "/")
+	if len(segments) == 0 {
+		return ctx
+	}
+
+	// For single-segment paths (e.g., "ipProtocol"), add flat key with variations
+	if len(segments) == 1 {
+		propName := segments[0]
+		ctx[propName] = value
+		addCfnNameVariations(ctx, propName, value)
+		return ctx
+	}
+
+	// For multi-segment paths, build nested structure
+	// Also add the leaf value directly for backward compatibility
+	leafName := segments[len(segments)-1]
+	ctx[leafName] = value
+	addCfnNameVariations(ctx, leafName, value)
+
+	// Build nested structure: fileSystemProtection/replicationOverwriteProtection
+	// becomes {"FileSystemProtection": {"ReplicationOverwriteProtection": value}}
+	buildNestedContext(ctx, segments, value)
+
+	return ctx
+}
+
+// buildNestedContext builds a nested object structure in the context from path segments.
+// For path ["fileSystemProtection", "replicationOverwriteProtection"] with value "DISABLED",
+// it adds multiple naming variations to ctx:
+//
+//	ctx["fileSystemProtection"] = {"replicationOverwriteProtection": "DISABLED"}
+//	ctx["FileSystemProtection"] = {"ReplicationOverwriteProtection": "DISABLED"}
+func buildNestedContext(ctx map[string]interface{}, segments []string, value interface{}) {
+	if len(segments) < 2 {
+		return
+	}
+
+	// Build the nested structure from innermost to outermost
+	// Start with the leaf value
+	current := value
+
+	// Work backwards through segments (except the first)
+	// to build the nested map structure
+	for i := len(segments) - 1; i >= 1; i-- {
+		propName := segments[i]
+		nested := make(map[string]interface{})
+		nested[propName] = current
+		addCfnNameVariations(nested, propName, current)
+		current = nested
+	}
+
+	// Add the outermost level to context with all name variations
+	rootName := segments[0]
+	ctx[rootName] = current
+	addCfnNameVariations(ctx, rootName, current)
+}
+
+// addCfnNameVariations adds multiple naming variations of a property to the context.
+// CloudFormation uses PascalCase with uppercase acronyms (DBClusterIdentifier, IpProtocol).
+func addCfnNameVariations(ctx map[string]interface{}, sdkName string, value interface{}) {
```

Comment:
I think we can simplify a lot here by passing through the `irreversibleNames` and using `naming.ToCfnName`. It would look something like this:

```go
func addCfnNameVariations(
      ctx map[string]interface{},
      sdkName string,
      value interface{},
      irreversibleNames map[string]string,
) {
    // Prefer CloudFormation naming derived from irreversible map when available.
    cfnName := naming.ToCfnName(sdkName, irreversibleNames)
    ctx[cfnName] = value
}
```

Then we can also get rid of `replaceCamelCaseAcronyms`, `replaceAcronymAtWordBoundary`, etc.